Elasticsearch: Add Support for Serial Differencing Pipeline Aggregation

simianhacker opened the pull request:
**What this PR does / why we need it**:

This PR closes #25621 by adding support for the serial differencing pipeline aggregation. Per the original feature request:

> To draw comparison graphs to detect trends and alert on if there is a change w.r.t for example last week. The built-in time shift feature in Grafana cannot operate on single metrics, and many backends (prometheus, influx) already support this by writing a suitable query.

![image](https://user-images.githubusercontent.com/41702/97496045-a359fc80-1925-11eb-8612-3b300dc8cccf.png)

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #25621

**Special notes for your reviewer**:

None

Elfo404 reviewed (changes requested):
Thanks for the PR! I tested it and it seems to work as expected. 

Would you mind adding a little test case for the query builder in `public/app/plugins/datasource/elasticsearch/specs/query_builder.test.ts` to check if the query gets built correctly?

Elfo404 commented:
We are planning to merge this as soon as the react migration (https://github.com/grafana/grafana/pull/28033) is complete.

Thanks for the contribution!

simianhacker committed:
Elasticsearch: Add support for serial diff pipeline aggregation

simianhacker commented on public/app/plugins/datasource/elasticsearch/types.ts:
```diff
@@ -32,6 +32,7 @@ interface MetricConfiguration<T extends MetricAggregationType> {
   hasSettings: boolean;
   hasMeta: boolean;
   defaults: Omit<Extract<MetricAggregation, { type: T }>, 'id' | 'type'>;
+  transform?: Record<string, (value: any) => any>;
```
@Elfo404 Is there a way to set the types up so the signature of `transform` matches the keys from the `settings` for an aggregation? OR would  we need to create 2 types of `MetricAggregationConfig`?

Elfo404 replied:
as far as I understood you need to transform what is saved in the query model (in this case `lag`) into a number because Elasticsearch expects it to be a number in the query, right?

while I think the solution is pretty neat (and I'm sure we can manage to make it type safe), this only works when the queries are issued by the frontend and won't support template variables. As we are planning to deprecate browser access mode (and therefore remove the query builder in a future release https://github.com/grafana/grafana/pull/29649) leveraging the plugin backend to build the query.

Given this, I had the idea of doing something similar by using a `format` function (like your `transform` one) in https://github.com/grafana/grafana/blob/82b21fe35e9bbab99bf3c185e17ab6834ef70776/public/app/plugins/datasource/elasticsearch/datasource.ts#L323  for all the `settings`, `meta` etc.. for each bucket and metric aggregation.

The formatting logic would have to be duplicated in the backend though as we would have to save the original string in the query model, which is less than ideal.

My suggestion is to drop that formatting logic for now, and simply treat `lag` as a number (like we already do for the MovingAverage Settings) while we find a better way to handle this in a more generic way, supporting template variables in the frontend and having a generic formatter in the backend, so instead of using `SettingField` something like: 

```tsx
<InlineField label="Lag">
  <Input
    onBlur={e => dispatch(changeMetricSetting(metric, 'lag', parseInt(e.target.value!, 10)))}
    defaultValue={metric.settings?.lag}
  />
</InlineField>
```

simianhacker replied:
Sounds good... I went ahead and implemented your suggestion and removed the `transform` stuff.

simianhacker committed:
Removing settings transsforms

simianhacker committed:
Removing unused deps

simianhacker committed:
removing unused dep

simianhacker commented:
@Elfo404 This has been converted and should be ready for review.

simianhacker committed:
Fixing type in test

marefr commented:
Does this work with Grafana Alerting? Otherwise backend code has to be implemented to support it. @Elfo404 have you checked?

Elfo404 commented:
@marefr didn't have time to properly test this yet, but thanks for the heads up

simianhacker committed:
Adding backend support for serial_diff

simianhacker commented:
@marefr I added the support back in for alerting (with a test). It was originally in there but when I rebased against the React changes it got dropped. Thanks for catching that 👍

Elfo404 reviewed (approved):
LGTM! 🚀

marefr commented:
@simianhacker Thank you for contributing to Grafana!
